plugin/hpsa/hpsa.py

```diff
@@ -918,3 +955,222 @@ def volume_ident_led_clear(self, volume, flags=Client.FLAG_RSVD):
+        else:
+            raise LsmError(ErrorNumber.PLUGIN_BUG,
+                           "Got unknown pdc: %d" % pdc)
+        self._sacli_exec(cmd, flag_force=True, flag_convert=False)
+
+    @_handle_errors
+    def volume_write_cache_policy_set(self, volume, wcp,
+                                      flags=Client.FLAG_RSVD):
+        """
+        Depending on these commands:
+            To disable both read and write cache:
+                hpssacli ctrl slot=3 ld 1 modify aa=disable
+            To enable no battery write cache:
+                hpssacli ctrl slot=0 modify nbwc=enable
+        """
+        (ctrl_num, array_num, ld_num) = self._cal_of_lsm_vol(volume)
+
+        cmd1 = ['ctrl', 'slot=%s' % ctrl_num, 'ld', ld_num]
+        cmd2 = []
+        if wcp == Volume.WRITE_CACHE_POLICY_WRITE_BACK:
+            cmd1.extend(['modify', 'aa=enable'])
+            cmd2 = ['ctrl', 'slot=%s' % ctrl_num, 'modify', 'nbwc=enable']
+        elif wcp == Volume.WRITE_CACHE_POLICY_AUTO:
+            cmd1.extend(['modify', 'aa=enable'])
+            cmd2 = ['ctrl', 'slot=%s' % ctrl_num, 'modify', 'nbwc=disable']
+        elif wcp == Volume.WRITE_CACHE_POLICY_WRITE_THROUGH:
+            cmd1.extend(['modify', 'aa=disable'])
+        else:
+            raise LsmError(ErrorNumber.PLUGIN_BUG,
+                           "Got unknown wcp: %d" % wcp)
+
+        try:
+            self._sacli_exec(cmd1, flag_force=True, flag_convert=False)
+            if cmd2:
+                self._sacli_exec(cmd2, flag_force=True, flag_convert=False)
+        except ExecError as exec_error:
+            # Check whether we got SSD volume
+            if self._is_ssd_volume(volume):
+                raise LsmError(
+                    ErrorNumber.NO_SUPPORT,
+                    "HP SmartArray does not allow changing SSD volume's "
+                    "cache policy while SmartPath is enabled")
+            raise exec_error
+
+    @_handle_errors
+    def volume_read_cache_policy_set(self, volume, rcp, flags=Client.FLAG_RSVD):
+        """
+        Depending on this command:
+            To disable both read and write cache:
+                hpssacli ctrl slot=3 ld 1 modify aa=disable
+        """
+        (ctrl_num, array_num, ld_num) = self._cal_of_lsm_vol(volume)
+
+        cmd = ['ctrl', 'slot=%s' % ctrl_num, 'ld', ld_num, 'modify']
+        if rcp == Volume.READ_CACHE_POLICY_DISABLED:
+            cmd.append('aa=disable')
+        elif rcp == Volume.READ_CACHE_POLICY_ENABLED:
+            cmd.append('aa=enable')
+        else:
+            raise LsmError(ErrorNumber.PLUGIN_BUG,
+                           "Got unknown rcp: %d" % rcp)
+        try:
+            self._sacli_exec(cmd, flag_force=True, flag_convert=False)
+        except ExecError as exec_error:
+            # Check whether we got SSD volume
+            if self._is_ssd_volume(volume):
+                raise LsmError(
+                    ErrorNumber.NO_SUPPORT,
```

Comment:
The write cache policy's error message is accurate, this error message is not. Also, keep in mind that we could (should?) add a method to disable Smart Path (but call it something generic, like "bypass" or something). Not sure that we'll get it in this release, but add it to a backlog somewhere.